src/core/src/engine/fork_resolver.rs

```diff
@@ -14,229 +14,302 @@
+                        error!("Error getting block from validator");
+                        break;
+                    }
+                }
+            }
         }
-    }
-}
 
-fn get_cert_from(block: &Block) -> WaitCertificate {
-    let payload = block.payload.clone();
-    debug!("Extracted payload from block: {:?}", payload.clone());
-    let (wait_certificate, _) = poet2_util::payload_to_wc_and_sig(&payload);
-    debug!("Serialized wait_cert : {:?}", &wait_certificate);
-    serde_json::from_str(&wait_certificate).unwrap()
+        // Loop over fork and chain to find a common ancestor
+        // Descend to the ancestor/previous block in each
+        // iteration for both the chain & the fork.
+        // If genesis is reached in the process( which is almost
+        // not possible), stop iteration and a completely new
+        // chain is competing for fork resolution.
+        loop {
+
+            let prev_chain_block_id = chain_block.previous_id.clone();
+            let prev_fork_block_id = fork_block.previous_id;
+            if prev_fork_block_id == prev_chain_block_id {
+                // Found common ancestor
+                info!("Found a common ancestor at block_id {}",
+                    poet2_util::to_hex_string(&prev_fork_block_id));
+                ancestor_found = true;
+                break;
+            }
+            let blocks_map = service
+                .get_blocks(vec![prev_chain_block_id.clone(),
+                             prev_fork_block_id.clone()]);
+            match blocks_map {
+                Err(err) => {
+                    error!("Could not get blocks");
+                    ancestor_found = false;
+                    break;
+                }
+                Ok(mut block_map) => {
+                    // Remove from the returned hashmap to get block
+                    chain_block = block_map.remove(&prev_chain_block_id)
+                        .expect("Could not extract block from map.");
+                    self.chain_len += 1;
+                    // Get cc from certificate in chain_block
+                    self.chain_cc += self.get_wait_time_from(&chain_block);
+
+                    match block_map.remove(&prev_fork_block_id){
+                         Some(mut prev_fork_block) => {
+                             debug!("Fork block num {} Chain block num {}", prev_fork_block.block_num, chain_block.block_num);
+                             fork_block = prev_fork_block;
+                             if fork_block.block_num == 0 {
+                                 warn!("Genesis reached while finding common ancestor.");
+                                 ancestor_found = true;
+                                 break;
+                             }
+                             // Keep adding wait times
+                             // Get cc from certificate in fork_block
+                             self.fork_cc += self.get_wait_time_from(&fork_block);
```

Comment:
extract wait time from wait certificate, store the result temporarily in a map and add it to state store if the fork wins